tensorflow/core/kernels/data/optional_ops_util.cc

```diff
@@ -0,0 +1,78 @@
+/* Copyright 2022 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#define EIGEN_USE_THREADS
+
+#include "tensorflow/core/kernels/data/optional_ops_util.h"
+
+#include "tensorflow/core/framework/op_kernel.h"
+
+namespace tensorflow {
+namespace data {
+
+Status OptionalZerosLike(OpKernelContext* ctx, const OptionalVariant& x,
+                         OptionalVariant* y,
+                         std::function<Status(OpKernelContext* ctx,
+                                              const Tensor& input, Tensor* out)>
+                             zeros_like_func) {
+  if (!x.has_value()) {
+    *y = x;
+    return OkStatus();
+  }
+  std::vector<Tensor> zero_tensors;
+  for (const Tensor& tensor : x.get_values()) {
+    Tensor zero_t;
+    TF_RETURN_IF_ERROR(zeros_like_func(ctx, tensor, &zero_t));
+    zero_tensors.push_back(std::move(zero_t));
+  }
+  *y = OptionalVariant(zero_tensors);
+  return OkStatus();
+}
+
+Status OptionalBinaryAdd(
+    OpKernelContext* ctx, const OptionalVariant& a, const OptionalVariant& b,
+    OptionalVariant* out,
+    std::function<Status(OpKernelContext* ctx, const Tensor& a, const Tensor& b,
+                         Tensor* out)>
+        binary_add_func) {
+  // TODO(skyewm): should adding a value to a non-value be a no-op instead?
+  if (a.has_value() != b.has_value()) {
+    return errors::InvalidArgument(
+        "Cannot add optionals because one has a value and the other doesn't.");
+  }
+  if (!a.has_value()) {
+    *out = a;
```

Comment:
[Internal mutation testing](https://testing.googleblog.com/2021/04/mutation-testing.html) detected that all existing tests would pass if this line is removed. Can you write a test that would check that `*y` gets the value of `x` in this scenario, so that accidentally removing this line would be caught by CI, please?

Reply:
It's true, this line isn't necessary because the caller functions already initialize the output to an empty optional beforehand. I can just remove the line.